csvcubed/csvcubed/readers/columnschema.py

```diff
@@ -0,0 +1,447 @@
+                )
+
+            elif isinstance(self, NewAttributeProperty):
+                return NewQbAttribute(
+                    label=self.label,
+                    description=self.description,
+                    new_attribute_values=_get_new_attribute_values(
+                        data, self.values
+                    ),
+                    parent_attribute_uri=self.from_existing,
+                    source_uri=self.definition_uri,
+                    is_required=self.required
+                )
+            else:
+                raise ValueError(f"Unhandled value: {self}")
+
+    def map_to_existing_qb_attribute(
+        self, data: PandasDataTypes
+    ) -> Union[ExistingQbAttribute, ExistingQbAttributeLiteral]:
+        if self.literalValuesDataType is None:
+            return ExistingQbAttribute(
+                self.uri,
+                new_attribute_values=_get_new_attribute_values(
+                    data, self.newAttributeValues
+                ),
+                is_required=self.isRequired,
+            )
+        else:
+            if self.newAttributeValues:
+                raise Exception(
+                    "Attributes cannot represent both literal values and attribute (resource) values"
+                )
+            return ExistingQbAttributeLiteral(
+                attribute_uri=self.uri,
+                data_type=self.data_type,
+                is_required=self.required,
+            )
+
+
+@dataclass
+class ExistingUnits(SchemaBaseClass):
+    value: str
+
+    def map_to_qb_multi_units(
+        self, data: PandasDataTypes, column_title: str
+    ) -> QbMultiUnits:
+        return QbMultiUnits.existing_units_from_data(
+            data, csvw_column_name_safe(column_title), self.value
+        )
+
+
+@dataclass
+class UnitType(SchemaBaseClass):
+    label: str
+    description: Optional[str] = None
+    from_existing: Optional[str] = None
+    definition_uri: Optional[str] = None
+    scaling_factor: Optional[float] = None
+    si_scaling_factor: Optional[float] = None
+    quantity_kind: Optional[str] = None
+
+
+@dataclass
+class Units(SchemaBaseClass):
+    values: Union[bool, List[UnitType]] = True
+
+    def map_to_qb_multi_units(self, data: PandasDataTypes) -> QbMultiUnits:
+        if isinstance(self.values, bool) and self.values is True:
+            return QbMultiUnits.new_units_from_data(data)
+
+        elif isinstance(self.values, list):
+
+            units = []
+            for unit in self.values:
+                if not isinstance(unit, UnitType):
+                    raise ValueError(f"Unexpected unit value: {unit}")
+
+                if any([unit.label, unit.description, unit.definition_uri]):
+                    # NewUnit
+                    units.append(_map_unit(unit))
+
+                elif unit.from_existing:
+                    units.append(ExistingQbUnit(unit_uri=self.values.from_existing))
```

Comment:
I don't think `self.values.from_existing` is going to resolve to anything since `self.values` is a list.